Move Bolt connector config to public config API

jakewins opened the pull request:
- Refactor the 'connector' configuration to achieve three things:
  - Change the API approach to a design that allows deducing the
    group configuration through introspection, allowing auto-docs
    for these settings
  - Move the Bolt connector configuration to GDS, because we want
    it to be publicly documented and because we want users of GDS
    to be able to enable Bolt in order to migrate existing GDS
    apps gradually.
  - Split the Connector group into a general `Connector` with
    the key, `enabled` and `type`. This allows Connector to be
    expanded as a general configuration API for configuring
    public ports (eg. have HTTP connector, Shell connector, etc etc).
    As per discussion with operability team (by which I mean I had
    a phone call with Ben)

jakewins committed:
Move Bolt connector config to public config API

- Refactor the 'connector' configuration to achieve three things:

  - Change the API approach to a design that allows deducing the
    group configuration through introspection, allowing auto-docs
    for these settings
  - Move the Bolt connector configuration to GDS, because we want
    it to be publicly documented and because we want users of GDS
    to be able to enable Bolt in order to migrate existing GDS
    apps gradually.
  - Split the Connector group into a general `Connector` with
    the key, `enabled` and `type`. This allows Connector to be
    expanded as a general configuration API for configuring
    public ports (eg. have HTTP connector, Shell connector, etc etc).
    As per discussion with operability team (by which I mean I had
    a phone call with Ben)

jakewins commented:
ping @benbc 

My thinking is we could then have other connector sub-classes, in the same way we have the `BoltConnector` one. I'm going to follow this PR (or amend it, depending on when it gets merged) to expand auto-documentation

benbc commented:
@jakewins Thanks for tackling this. I _really_ like what you've done. I have one doubt, though: are you really going to be able to get the doc generation to work? My worry is that the generator won't be able to statically extract all the information it needs. It can just about get from the `@Description` decoration on the `Connector` class to the individual subclasses by following the inheritance hierarchy (although even that is rather non-local for the general case where connectors are defined in multiple settings files) but how can it extract the defaults? There needs to be a static definition of the individual settings rather than runtime definition as at the moment.

jakewins committed:
Refactor config asciidoc generator.

- Partially to clean it up and add rudimentary testing,
  but mainly to help myself understand it in order to add support
  for documenting grouped settings.
- Add auto-docs for grouped configuration

jakewins committed:
Add Embedded example for configuring Bolt

jakewins commented:
Merging as per sync with @benbc
